cfme/common/provider.py

```diff
@@ -742,39 +687,33 @@ def _fill_credential(form, cred, validate=None):
                 # ``exc`` must have some contents by now since at least one except had to happen.
                 raise exc
 
-    elif cred.type == 'azure':
+    elif isinstance(cred, AzureCredential):
         fill(cred.form, {'default_username': cred.principal,
                          'default_password': cred.secret,
                          'default_verify': cred.secret})
-    elif cred.type == 'ssh':
+    elif isinstance(cred, SSHCredential):
         fill(cred.form, {'ssh_user': cred.principal, 'ssh_key': cred.secret})
-    elif cred.type == 'token':
-        if version.current_version() < "5.6":
-            fill(cred.form, {'token_secret_55': cred.token, 'validate_btn': validate})
-        else:
+    elif isinstance(cred, TokenCredential):
+        fill(cred.form, {
+            'token_secret': cred.token,
+            'token_verify_secret': cred.verify_token,
+            'validate_btn': validate
+        })
+        if validate:
+            # Validate default creds and move on to hawkular tab validation
+            flash.assert_no_errors()
             fill(cred.form, {
-                'token_secret': cred.token,
-                'token_verify_secret': cred.verify_token,
-                'validate_btn': validate
+                'hawkular_validate_btn': validate
             })
-            if validate:
-                # Validate default creds and move on to hawkular tab validation
-                flash.assert_no_errors()
-                fill(cred.form, {
-                    'hawkular_validate_btn': validate
-                })
-    elif cred.type == 'service_account':
+    elif isinstance(cred, ServiceAccountCredential):
         fill(cred.form, {'google_service_account': cred.service_account, 'validate_btn': validate})
     else:
-        if cred.domain:
-            principal = r'{}\{}'.format(cred.domain, cred.principal)
-        else:
-            principal = cred.principal
-        fill(cred.form, {'default_principal': principal,
+        fill(cred.form, {'default_principal': cred.principal,
```

Comment:
while reading backwards it was hard to notice that this logic got pushed into the magic attribute